include/benchmark/benchmark_api.h

```diff
@@ -331,16 +344,19 @@ class State {
   // Returns true if the benchmark should continue through another iteration.
   // NOTE: A benchmark may not return from the test until KeepRunning() has
   // returned false.
-  bool KeepRunning() {
-    if (BENCHMARK_BUILTIN_EXPECT(!started_, false)) {
-      StartKeepRunning();
-    }
-    bool const res = total_iterations_++ < max_iterations;
-    if (BENCHMARK_BUILTIN_EXPECT(!res, false)) {
-      FinishKeepRunning();
-    }
-    return res;
-  }
+  bool KeepRunning();
+
+  // Returns true if the benchmark should process another 'n' iterations. The
```

Comment:
Thanks for bringing this up. It does sound like probably a cleaner solution. There is one situation where I'm seeing it complicate things: 

As much as I'd like it not to be the case, I think the range-based for loop API will depend on this behavior somewhat. Especially if we want to document it as the API of choice. The range-based for uses 1000 as the batch size, but the API could easily be used by a benchmark which is slow enough to only run in the single digits to tens of iterations in a reasonable amount of time. So we want to easily be able to cut the batch size short when that feature comes.

I guess thinking more, we could give a bool flag for State for whether KeepRunningBatch was called or not and use that when calculating the number of iterations. (Also would add the batch size as a State attribute and update the contract that the batch size must remain the same)

Thanks again for bringing this up. I will work on this tomorrow, but I'll leave this comment here already in case you have some thoughts to add in the meantime.